### ISSUE TITLE
numpy.linalg.eigh fails to properly diagonalize some matrices while raising no error

### ISSUE THREAD

**Issue by DAGwood1018** [numpy/numpy#19472]
While diagonalizing the blocks of a large matrix I encountered several Hermitian blocks that numpy.linalg.eigh does not appear to be able to handle. By this I mean the eigenvalues it returns do not add up to the trace of the matrix indicating they must be wrong. 

### Reproducing code example:

Example of a matrix that causes eigh fails to diagonalize correctly is contained in zip file:
[Matrix.zip](https://github.com/numpy/numpy/files/6813181/Matrix.zip)

```python
import numpy as np
import os

path= os.getcwd()
with open(os.path.join(path,'mat.npy'),'rb') as file:
    M= np.load(file, allow_pickle=False)

w, v= np.linalg.eigh(M)
    
trace= np.trace(M)
eigSum= np.sum(w)
if not round(trace,7)==round(eigSum,7):
    print('Diagonalization Failed?')
    print('Trace=', trace)
    print('Eigval Sum=', eigSum)
```

### Error message:
Numpy does not raise any error message but it is clear that the diagonalization cannot be correct since the trace does not equal the sum of the eigenvalues. 

### NumPy/Python version information:
Python 3.8
Numpy 1.19.2

**Comment by bashtage** [comment]
Are you shure `eigh` is appropriate?  M does not appear to be symmetric

```
In [8]: np.abs((M+M.T)).sum()

Out[8]: 3245760.0
```

**Comment by DAGwood1018** [comment]
Yeah, I checked if M was Hermitian/symmetric by 

`np.sum(np.subtract(M,M.H))`

which gave 0 as expected for such a matrix. I also tried 

`if not np.isclose(M,M.H):
       print('M is not Hermitian')`

which was never triggered. 

If it helps I did use scipy.sparse.linalg.eigsh() to check for the magnitudes of a few eigenvalues and this function appeared to work as it gave reasonable eigenvalues. The eigenvalues appeared to be on the interval of ~(-40) to ~(-0.8).

**Comment by Dinkarkumar** [comment]
Is the issue still open ? Can I work on this issue ?

**Comment by DAGwood1018** [comment]
> 
> 
> Is the issue still open ? Can I work on this issue ? 

I believe numpy eigh still doesn't work on the matrix I provided. I did find that scipy's implementation of eigh does work (and have just been using that) so I imagine one can figure out what is going wrong by cross referencing them.

**Comment by Dinkarkumar** [comment]
I am going through the issue, would seek your insight whenever i need, since i am new to the opensource contribution. Thank you for your reply.

**Comment by Dinkarkumar** [comment]
Matrix.npy file is around 7.24GB, is it really that large or the file is corrupted ?

**Comment by DAGwood1018** [comment]
> 
> 
> Matrix.npy file is around 7.24GB, is it really that large or the file is corrupted ?

No, the problem only occurs for very large matrices.

**Comment by charris** [comment]
Can you try the nightly builds, `pip install -i https://pypi.anaconda.org/scipy-wheels-nightly/simple numpy`. They are built with 64 bit BLAS integers and more recent OpenBLAS.

**Comment by Dinkarkumar** [comment]
With Scipy, its taking around 10hrs(and counting) to calculate the eigen values of the matrix u have given.

**Comment by DAGwood1018** [comment]
Yeah, its a really large matrix that takes quite some time. I actually plan on just performing the diagonalization via LAPACK directly for a faster runtime. In other words, this I wouldn't call this a critically pressing issue as it seem to only appear for some large matrices where using C/Fortran are the better options. I just found it strange that numpy's implementation fails while scipy's doesn't despite both relying on LAPACK
